File: bindings/swift/Sources/MaplibreNativeSupport/QueryStructs.swift
import CMaplibreNativeC

public enum NativeRenderedQueryGeometry: Equatable, Sendable {
  case point(NativeScreenPoint)
  case box(min: NativeScreenPoint, max: NativeScreenPoint)
  case lineString([NativeScreenPoint])

  public func withNativeGeometry<Result>(
    _ body: (UnsafePointer<mln_rendered_query_geometry>) throws -> Result
  ) throws -> Result {
    switch self {
    case .point(let point):
      var geometry = mln_rendered_query_geometry_point(point.native)
      return try withUnsafePointer(to: &geometry, body)
    case .box(let min, let max):
      var geometry = mln_rendered_query_geometry_box(mln_screen_box(min: min.native, max: max.native))
      return try withUnsafePointer(to: &geometry, body)
    case .lineString(let points):
      let nativePoints = points.map(\.native)
      return try nativePoints.withUnsafeBufferPointer { buffer in
        var geometry = mln_rendered_query_geometry_line_string(buffer.baseAddress, buffer.count)
        return try withUnsafePointer(to: &geometry, body)
      }
    }
  }
}

public struct NativeRenderedFeatureQueryOptions: Equatable, Sendable {
  public let layerIds: [String]
  public let filter: NativeJSONValue?

  public init(layerIds: [String] = [], filter: NativeJSONValue? = nil) {
    self.layerIds = layerIds
    self.filter = filter
  }

  public func withNativeOptions<Result>(
    _ body: (UnsafePointer<mln_rendered_feature_query_options>?) throws -> Result
  ) throws -> Result {
    if layerIds.isEmpty, filter == nil { return try body(nil) }
    let arena = NativeJSONArena()
    let layerViews = layerIds.map { arena.view($0) }
    return try layerViews.withUnsafeBufferPointer { layerViews in
      var options = mln_rendered_feature_query_options_default()
      if !self.layerIds.isEmpty {
        options.fields |= MLN_RENDERED_FEATURE_QUERY_OPTION_LAYER_IDS.rawValue
        options.layer_ids = layerViews.baseAddress
        options.layer_id_count = layerViews.count
      }
      if let filter {
        options.filter = arena.allocate(filter)
      }
      return try withUnsafePointer(to: &options, body)
    }
  }
}

public struct NativeSourceFeatureQueryOptions: Equatable, Sendable {
  public let sourceLayerIds: [String]
  public let filter: NativeJSONValue?

  public init(sourceLayerIds: [String] = [], filter: NativeJSONValue? = nil) {
    self.sourceLayerIds = sourceLayerIds
    self.filter = filter
  }

  public func withNativeOptions<Result>(
    _ body: (UnsafePointer<mln_source_feature_query_options>?) throws -> Result
  ) throws -> Result {
    if sourceLayerIds.isEmpty, filter == nil { return try body(nil) }
    let arena = NativeJSONArena()
    let layerViews = sourceLayerIds.map { arena.view($0) }
    return try layerViews.withUnsafeBufferPointer { layerViews in
      var options = mln_source_feature_query_options_default()
      if !self.sourceLayerIds.isEmpty {
        options.fields |= MLN_SOURCE_FEATURE_QUERY_OPTION_SOURCE_LAYER_IDS.rawValue
        options.source_layer_ids = layerViews.baseAddress
        options.source_layer_id_count = layerViews.count
      }
      if let filter {
        options.filter = arena.allocate(filter)
      }
      return try withUnsafePointer(to: &options, body)
    }
  }
}

public struct NativeFeatureQueryResultReader {
  public let handle: OpaquePointer

  public init(handle: OpaquePointer) {
    self.handle = handle
  }

  public func copyFeatures() throws -> [NativeQueriedFeature] {
    let count = try CAPI.featureQueryResultCount(handle)
    return try (0..<count).map { index in
      try CAPI.featureQueryResultGet(handle, index: index)
    }
  }
}

File: bindings/dotnet/src/Maplibre.NativeFfi/Internal/Struct/QueryStructs.cs
using System.Runtime.InteropServices;
using Maplibre.NativeFfi.Internal.C;
using Maplibre.NativeFfi.Internal.Memory;
using Maplibre.NativeFfi.Query;

namespace Maplibre.NativeFfi.Internal.Struct;

internal sealed unsafe class NativeRenderedQueryGeometry : IDisposable
{
    private nint points;

    private NativeRenderedQueryGeometry(mln_rendered_query_geometry value, nint points)
    {
        Value = value;
        this.points = points;
    }

    internal mln_rendered_query_geometry Value { get; }

    internal static NativeRenderedQueryGeometry From(RenderedQueryGeometry geometry)
    {
        ArgumentNullException.ThrowIfNull(geometry);
        return geometry switch
        {
            RenderedQueryGeometry.Point point => new(
                NativeMethods.mln_rendered_query_geometry_point(MapStructs.ToNative(point.Value)),
                0
            ),
            RenderedQueryGeometry.Box box => new(
                NativeMethods.mln_rendered_query_geometry_box(
                    new mln_screen_box
                    {
                        min = MapStructs.ToNative(box.Value.Min),
                        max = MapStructs.ToNative(box.Value.Max),
                    }
                ),
                0
            ),
            RenderedQueryGeometry.LineString line => FromLineString(line),
            _ => throw new ArgumentException(
                $"Unsupported rendered query geometry type {geometry.GetType().Name}.",
                nameof(geometry)
            ),
        };
    }

    private static NativeRenderedQueryGeometry FromLineString(RenderedQueryGeometry.LineString line)
    {
        var nativePoints = NativeAllocation.AllocArray<mln_screen_point>(line.Points.Count);
        for (var index = 0; index < line.Points.Count; index++)
        {
            nativePoints[index] = MapStructs.ToNative(line.Points[index]);
        }
        return new NativeRenderedQueryGeometry(
            NativeMethods.mln_rendered_query_geometry_line_string(
                nativePoints,
                (nuint)line.Points.Count
            ),
            (nint)nativePoints
        );
    }

    public void Dispose()
    {
        if (points != 0)
        {
            NativeMemory.Free((void*)points);
            points = 0;
        }
    }
}

internal sealed unsafe class NativeRenderedFeatureQueryOptions : IDisposable
{
    private readonly List<NativeStringView> layerIds = [];
    private readonly NativeStringView? filter;
    private nint layerIdArray;

    private NativeRenderedFeatureQueryOptions(
        mln_rendered_feature_query_options value,
        NativeStringView? filter
    )
    {
        Value = value;
        this.filter = filter;
    }

    internal mln_rendered_feature_query_options Value { get; private set; }

    internal static NativeRenderedFeatureQueryOptions From(RenderedFeatureQueryOptions options)
    {
        ArgumentNullException.ThrowIfNull(options);
        var filter = options.Filter is null
            ? null
            : NativeStringView.From(options.Filter, nameof(options.Filter));
        var seeded = NativeMethods.mln_rendered_feature_query_options_default();
        seeded.filter = filter?.Pointer;
        var native = new NativeRenderedFeatureQueryOptions(seeded, filter);
        try
        {
            if (options.LayerIds is { } ids)
            {
                var value = native.Value;
                value.fields |= (uint)
                    mln_rendered_feature_query_option_field.MLN_RENDERED_FEATURE_QUERY_OPTION_LAYER_IDS;
                value.layer_id_count = (nuint)ids.Count;
                if (ids.Count > 0)
                {
                    var pointer = NativeAllocation.AllocArray<mln_buffer_view>(ids.Count);
                    native.layerIdArray = (nint)pointer;
                    for (var index = 0; index < ids.Count; index++)
                    {
                        var view = NativeStringView.From(ids[index], $"LayerIds[{index}]");
                        native.layerIds.Add(view);
                        pointer[index] = view.Value;
                    }
                    value.layer_ids = pointer;
                }
                native.Value = value;
            }
            return native;
        }
        catch
        {
            native.Dispose();
            throw;
        }
    }

    public void Dispose()
    {
        foreach (var value in layerIds)
            value.Dispose();
        if (layerIdArray != 0)
            NativeMemory.Free((void*)layerIdArray);
        filter?.Dispose();
    }
}

internal sealed unsafe class NativeSourceFeatureQueryOptions : IDisposable
{
    private readonly List<NativeStringView> sourceLayerIds = [];
    private readonly NativeStringView? filter;
    private nint sourceLayerIdArray;

    private NativeSourceFeatureQueryOptions(
        mln_source_feature_query_options value,
        NativeStringView? filter
    )
    {
        Value = value;
        this.filter = filter;
    }

    internal mln_source_feature_query_options Value { get; private set; }

    internal static NativeSourceFeatureQueryOptions From(SourceFeatureQueryOptions options)
    {
        ArgumentNullException.ThrowIfNull(options);
        var filter = options.Filter is null
            ? null
            : NativeStringView.From(options.Filter, nameof(options.Filter));
        var seeded = NativeMethods.mln_source_feature_query_options_default();
        seeded.filter = filter?.Pointer;
        var native = new NativeSourceFeatureQueryOptions(seeded, filter);
        try
        {
            if (options.SourceLayerIds is { } ids)
            {
                var value = native.Value;
                value.fields |= (uint)
                    mln_source_feature_query_option_field.MLN_SOURCE_FEATURE_QUERY_OPTION_SOURCE_LAYER_IDS;
                value.source_layer_id_count = (nuint)ids.Count;
                if (ids.Count > 0)
                {
                    var pointer = NativeAllocation.AllocArray<mln_buffer_view>(ids.Count);
                    native.sourceLayerIdArray = (nint)pointer;
                    for (var index = 0; index < ids.Count; index++)
                    {
                        var view = NativeStringView.From(ids[index], $"SourceLayerIds[{index}]");
                        native.sourceLayerIds.Add(view);
                        pointer[index] = view.Value;
                    }
                    value.source_layer_ids = pointer;
                }
                native.Value = value;
            }
            return native;
        }
        catch
        {
            native.Dispose();
            throw;
        }
    }

    public void Dispose()
    {
        foreach (var value in sourceLayerIds)
            value.Dispose();
        if (sourceLayerIdArray != 0)
            NativeMemory.Free((void*)sourceLayerIdArray);
        filter?.Dispose();
    }
}
